cql3/restrictions/statement_restrictions.cc

```diff
@@ -1406,8 +1398,160 @@ bool is_satisfied_by(
         }, restr);
 }
 
+/// A column's bound, from WHERE restrictions.
+class bound_t {
+    bool _unbounded;
+    bytes_opt _value; // Invalid when _unbounded is true.
+    const abstract_type& _value_type;
+public:
+    explicit bound_t(const abstract_type& t) : _unbounded(true), _value_type(t) {}
+    bound_t(const abstract_type& t, const bytes_opt& v) : _unbounded(false), _value(v), _value_type(t) {}
+
+    /// True iff *this is a tighter lower bound than \p that.
+    bool is_tighter_lb_than(const bound_t& that) const {
+        if (auto sc = shortcircuit(that)) {
+            return *sc;
+        } else {
+            // *this is a tighter lower bound if it's larger.
+            return _value_type.compare(*this->_value, *that._value) > 0;
+        }
+    }
+
+    /// True iff *this is a tighter upper bound than \p that.
+    bool is_tighter_ub_than(const bound_t& that) const {
+        if (auto sc = shortcircuit(that)) {
+            return *sc;
+        } else {
+            // *this is a tighter upper bound if it's smaller.
+            return _value_type.compare(*this->_value, *that._value) < 0;
+        }
+    }
+
+    bytes_view_opt value() const {
+        if (_unbounded) {
+            throw std::logic_error("bound_t::value() called on unbounded");
+        }
+        return _value;
+    }
+
+private:
+    /// If the comparison *this<=>that can be shortcircuited (due to unbounded or null cases), returns the
+    /// comparison result.  Otherwise, returns nullopt.
+    std::optional<bool> shortcircuit(const bound_t& that) const {
+        if (that._unbounded) {
+            return true; // Anything is tighter than no bound.
+        } else if (this->_unbounded || !that._value) { // null always wins: it makes the slice empty.
+            return false;
+        } else if (!this->_value) { // Now that we're sure _value is valid, check it for null.
+            return true;
+        } else {
+            return std::nullopt;
+        }
+    }
+};
+
+/// True iff op means bnd type of bound.
+bool matches(const operator_type* op, statements::bound bnd) {
+    static const std::vector<std::vector<const operator_type*>> operators{
+        {&operator_type::EQ, &operator_type::GT, &operator_type::GTE}, // These mean a lower bound.
+        {&operator_type::EQ, &operator_type::LT, &operator_type::LTE}, // These mean an upper bound.
+    };
+    const auto zero_if_lower_one_if_upper = get_idx(bnd);
+    return boost::algorithm::any_of_equal(operators[zero_if_lower_one_if_upper], op);
+}
+
+bound_t get_bound(const expression& restr, const query_options& options, statements::bound bnd) {
+    return std::visit(overloaded_functor{
+            [&] (bool v) {
+                if (v) {
+                    return bound_t(*byte_type);
+                } else {
+                    throw std::logic_error("empty set has no bounds");
```

Comment:
This is fishy but I don't think it's reachable now. Isn't it equivalent to `WHERE TRUE` and `WHERE FALSE`?

Reply:
This particular line (as opposed to 1467) is equivalent to `WHERE FALSE`.  Even if that were valid CQL, I can't think of a good value for get_bound() to return in that case.

This line *is* reachable, though, if a restriction's expression field was not set, because the default constructor just sets it to `false`.  It throws to make sure that I properly set that field for all restrictions that do trigger a call to get_bound().

Reply:
Shouldn't the default be `WHERE TRUE`?

Reply:
It should definitely default to TRUE after we drop the old code.  I just thought FALSE would be more useful while I'm hunting for discrepancies between old and new.